### PR TITLE
Polymorphic types.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Max!

For after https://github.com/c5t/Current/pull/273.

Polymorphic types, all but serialization by now.

**Suggestion**: Looks like with a bit of tweaking, the dream may come true. Specifically:

* Use global `Exists(...)` or `Exists<T>(...)` for base types, optionals, and polymorphics.
* Use global `Value(...)` or `Value<T>(...)` for base types, optionals, and polymorphics.

We'll have to be careful with type hierarchies of types within `Polymorphic<>`. As of now, `Match()` would do the right thing with RTTI dispatching, while `Value<T>` would happily case derived to base if requested. The above does sound right to me, but want to double-check with you.

Thanks,
Dima

PS: Better naming suggestions for `Match()`? I don't like it. `MatchWith()` is more functional, but still doesn't sound right in C++.

### CODE REVIEW

File: TypeSystem/Serialization/json.h [hunk 1]
```diff
@@ -212,160 +231,185 @@ struct LoadFromJSONImpl {
...
   }
 };
 
 template <typename T>
 struct LoadFromJSONImpl<std::vector<T>> {
-  static void Load(rapidjson::Value& source, std::vector<T>& destination, const std::string& path) {
-    if (!source.IsArray()) {
+  static void Load(rapidjson::Value* source, std::vector<T>& destination, const std::string& path) {
+    if (source && source->IsArray()) {
+      const size_t n = source->Size();
```
Comment (mzhurovich) [inline TypeSystem/Serialization/json.h]: `n`->`size`?

File: TypeSystem/polymorphic.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+template <typename T, typename... TS>
+struct PolymorphicImpl<T, TS...> {
+  using T_TYPELIST = TypeListImpl<T, TS...>;
+  enum { T_TYPELIST_SIZE = TypeListSize<T_TYPELIST>::value };
+
+  std::unique_ptr<CurrentSuper> object_;
+
+  // Initializes an `std::unique_ptr<CurrentSuper>` given the object of the right type.
+  // The input object could be an object itself (in which case it's copied),
+  // an `std::move()`-d `std::unique_ptr` to that object (in which cast it's moved),
```
Comment (mzhurovich) [inline TypeSystem/polymorphic.h]: s/cast/case/

File: TypeSystem/polymorphic.h [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+  // or a bare pointer (in which case it's captured).
+  // TODO(dkorolev): The bare pointer one is sure unsafe -- consult with @mzhurovich.
+  struct TypeCheckedAsisgnment {
```
Comment (mzhurovich) [inline TypeSystem/polymorphic.h]: s/Asisgnment/Assignment/

File: TypeSystem/Serialization/json.h (hunk 1 above)
Comment (dkorolev) [inline TypeSystem/Serialization/json.h]: :+1: 90329ba

File: TypeSystem/polymorphic.h (hunk 2 above)
Comment (dkorolev) [inline TypeSystem/polymorphic.h]: :+1: 90329ba

File: TypeSystem/polymorphic.h (hunk 3 above)
Comment (dkorolev) [inline TypeSystem/polymorphic.h]: :+1: 90329ba

**Comment by mzhurovich** [comment]
@dkorolev As for `Match()` - how about `Apply()` or `Call()` or `RTTICall()`?

**Comment by dkorolev** [comment]
Renamed `Match()` into `Call()`, and removed that confusing `const` in the constructor parameter to `Optional<>`.

@mzhurovich PTAL.

Thanks,
Dima

File: TypeSystem/polymorphic.h [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    template <typename Q>
+    static void Perform(Q&& source, std::unique_ptr<CurrentSuper>& destination) {
+      Instance instance;
+      instance(DerivedTypesDifferentiator<bricks::decay<Q>>(), std::forward<Q>(source), destination);
+    }
+  };
+
+  PolymorphicImpl() = delete;
+
+  // PolymorphicImpl(std::unique_ptr<CurrentSuper>&& rhs) : object_(std::move(rhs)) {}
```
Comment (mzhurovich) [inline TypeSystem/polymorphic.h]: Is it deliberately commented out?

Comment (dkorolev) [inline TypeSystem/polymorphic.h]: Fixed in 33fc228 !

**Comment by mzhurovich** [comment]
@dkorolev Помолясь применимъ.

**Comment by dkorolev** [comment]
Воистинъ Аминъ.

**Pull request merged**